docs/language-server/protocol-language-server.md

```diff
@@ -3918,6 +3964,349 @@ standard input.
 null;
 ```
 
+## Library-Related Operations
+
+The library-related operations provide the Language Server with capabilities to
+check and modify project's edition settings, list editions published in a given
+edition, create local library projects which can be imported in the currently
+opened project and publish them.
+
+### `editions/listAvailable`
+
+Lists editions available on the system.
+
+Moreover, if `update` is set to `true`, it will download any new editions from
+the repositories and include them in the result as well.
+
+> Currently, if `update` was `true` but some downloads failed, the endpoint will
+> still return a success, just containing the editions that were already
+> available. In the future it should emit warnings using proper notification
+> channels.
+
+#### Parameters
+
+```typescript
+{
+  update: Boolean;
+}
+```
+
+#### Result
+
+```typescript
+{
+  editionNames: [String];
+}
+```
+
+### `editions/resolve`
+
+Resolves settings implied by the edition.
+
+> Currently, it only resolves the engine version, as only it is needed, but
+> other settings may be added if necesssary.
+
+#### Parameters
+
+```typescript
+{
+    editionName?: String;
+}
+```
+
+The `editionName` is optional, if it is not provided, the (unnamed) edition from
+currently open project's settings is resolved, taking into account both its
+parent edition and any locally set overrides.
+
+#### Result
+
+```typescript
+{
+  engineVersion: String;
+}
+```
+
+#### Errors
+
+- [`EditionNotFoundError`](#editionnotfounderror) indicates that the requested
+  edition, or some edition referenced in the ancestors of the edition being
+  resolved, could not be found.
+- [`FileSystemError`](#filesystemerror) to signal a generic, unrecoverable
+  file-system error.
+
+### `editions/getProjectSettings`
+
+Returns the currently set edition-related settings of the project.
+
+Currently, it only returns the parent edition and local library preference. Once
+more advanced edition settings are to be supported in the IDE, this endpoint
+will be extended.
+
+#### Parameters
+
+```typescript
+null;
+```
+
+#### Result
+
+```typescript
+{
+    parentEdition?: String;
+    preferLocalLibraries: Boolean;
+}
+```
+
+The `parentEdition` may be missing if it is not set. It is possible to manually
+edit the `pacakge.yaml` and generate a valid edition config that does not
+specify a parent edition.
+
+### `editions/setProjectParentEdition`
+
+Sets the parent edition of the project to a specific edition.
+
+This change may even change the version of the engine associated with the
+project, so for the changes to take effect, the language server may need to be
+restarted. The endpoint only modifies the `pacakge.yaml` file, which is
+preloaded in the Language Server, so it is IDE's responsibility to re-open the
+project.
+
+#### Parameters
+
+```typescript
+{
+  newEditionName: String;
+}
+```
+
+#### Result
+
+```typescript
+null;
+```
```

Comment:
Would it make sense if this returns whether the engine needs to be restarted for changes to be applied? This would make it easier for the IDE to know whether it needs to take this action, and in the future if this no longer is required to do the restart, we would only update this response instead of having to update the logic in the IDE.

Reply:
Currently it would always return 'needs restart' and I'm not sure if changing edition is ever going to not need it (as it will need to reload almost all modules), but I guess we can add this response for possible forwards compatibility.

Reply:
Should the `setProjectLocalLibrariesPreference` endpoint also return a result like this? Here the story is different because for now it would always return `needsRestart: true` but if the 'hot-reload' is implemented at some point, it will start to always return `false`. But maybe it is worth for compatibility and once it is not needed anymore it can be removed from the protocol.

Reply:
Now I think of it - maybe I could make them optional - if they are not present, defaulting to `needsRestart: false` - this way we will be able to safely remove them from the protocol once that is implemented.

Reply:
> Should the setProjectLocalLibrariesPreference endpoint also return a result like this?

I'd say yes.

> Now I think of it - maybe I could make them optional - if they are not present, defaulting to needsRestart: false - this way we will be able to safely remove them from the protocol once that is implemented.

Sounds reasonable.